src/output-filestore.c

```diff
@@ -0,0 +1,482 @@
+                                                        * open
+                                                        * files */
+
+typedef struct OutputFilestoreCtx_ {
+    char prefix[FILESTORE_PREFIX_MAX];
+    char tmpdir[FILESTORE_PREFIX_MAX];
+    bool fileinfo;
+} OutputFilestoreCtx;
+
+typedef struct OutputFilestoreLogThread_ {
+    OutputFilestoreCtx *ctx;
+    uint32_t file_cnt;
+    uint16_t counter_max_hits;
+} OutputFilestoreLogThread;
+
+static uint64_t OutputFilestoreOpenFilesCounter(void)
+{
+    return SC_ATOMIC_GET(filestore_open_file_cnt);
+}
+
+static uint32_t g_file_store_max_open_files = 0;
+
+static void FileSetMaxOpenFiles(uint32_t count)
+{
+    g_file_store_max_open_files = count;
+}
+
+static uint32_t FileGetMaxOpenFiles(void)
+{
+    return g_file_store_max_open_files;
+}
+
+static void PrintHexString(char *str, size_t size, uint8_t *buf, size_t buf_len)
+{
+    int i = 0;
+    size_t x = 0;
+    for (i = 0, x = 0; x < buf_len; x++) {
+        i += snprintf(&str[i], size - i, "%02x", buf[x]);
+    }
+}
+
+/**
+ * \brief Update the timestamps on a file to match those of another
+ *     file.
+ *
+ * \param src_filename Filename to use as timestamp source.
+ * \param filename Filename to apply timestamps to.
+ */
+static void OutputFilestoreUpdateFileTime(const char *src_filename,
+        const char *filename)
+{
+    struct stat sb;
+    if (stat(src_filename, &sb) != 0) {
+        SCLogDebug("Failed to stat %s: %s", filename, strerror(errno));
+        return;
+    }
+    struct utimbuf utimbuf = {
+        .actime = sb.st_atime,
+        .modtime = sb.st_mtime,
+    };
+    if (utime(filename, &utimbuf) != 0) {
+        SCLogDebug("Failed to update file timestamps: %s: %s", filename,
+                strerror(errno));
+    }
+}
+
+static void OutputFilestoreFinalizeFiles(const OutputFilestoreCtx *ctx,
+        const Packet *p, File *ff) {
+    /* Stringify the SHA256 which will be used in the final
+     * filename. */
+    char sha256string[(SHA256_LENGTH * 2) + 1];
+    PrintHexString(sha256string, sizeof(sha256string), ff->sha256,
+            sizeof(ff->sha256));
+
+    char tmp_filename[PATH_MAX] = "";
+    snprintf(tmp_filename, sizeof(tmp_filename), "%s/file.%u", ctx->tmpdir,
+            ff->file_store_id);
+
+    char final_filename[PATH_MAX] = "";
+    snprintf(final_filename, sizeof(final_filename), "%s/%c%c/%s",
+            ctx->prefix, sha256string[0], sha256string[1], sha256string);
+
+    if (SCPathExists(final_filename)) {
+        OutputFilestoreUpdateFileTime(tmp_filename, final_filename);
+        if (unlink(tmp_filename) != 0) {
+            SCLogWarning(SC_WARN_REMOVE_FILE,
+                    "Failed to remove temporary file %s: %s", tmp_filename,
+                    strerror(errno));
+        }
+    } else if (rename(tmp_filename, final_filename) != 0) {
+        SCLogWarning(SC_WARN_RENAMING_FILE, "Failed to rename %s to %s: %s",
```

Comment:
should we remove the failed file here?

I guess there are several possible scenarios:
1. some (advanced) user puts tmp on a different mount, and while moving the target has no space
2. permissions of target dir are bad. We can probably check this at startup?

Btw (1) above is probably very bad wrt perf as it is going to wait for expensive fs ops (copy of potentially large files)

Reply:
We should probably remove it. Its likely that writing metadata will fail to then, and we don't want to leave too many hanging files.

Reply:
Fixed. Attempt to unlink on rename fail.